File: fpcparser.py
import antlr4
from gen.FPCoreLexer import FPCoreLexer
from gen.FPCoreParser import FPCoreParser
from gen.FPCoreVisitor import FPCoreVisitor

import fpcast as ast
ops = ast.operations

def propToSexp(p):
    if isinstance(p, str):
        return '"' + p + '"'
    elif isinstance(p, list):
        return '(' + ' '.join(p) + ')'
    else:
        return str(p)

class FPCoreObject(object):
    def __init__(self, input_vars, props, e):
        self.args = input_vars
        self.properties = props
        self.name = self.properties.get(':name', None)
        self.pre = self.properties.get(':pre', None)
        self.e = e

    def __str__(self):
        return 'FPCore ({})\n  name: {}\n   pre: {}\n  {}'.format(
            ' '.join(self.args), self.name, self.pre, self.e)

    def __repr__(self):
        return 'FPCoreObject(\n  {},\n  {},\n  {}\n)'.format(
            repr(self.args), repr(self.properties), repr(self.e))

    def _sexp(self):
        return '(FPCore ({}) {} {})'.format(
            ' '.join(self.args),
            ' '.join(name + ' ' + propToSexp(prop) for name, prop in self.properties.items()),
            str(self.e))
    sexp = property(_sexp)

class Visitor(FPCoreVisitor):
    def visitParse(self, ctx):
        cores = []
        for child in ctx.getChildren():
            parsed = child.accept(self)
            if parsed:
                cores.append(parsed)
        return cores

    def visitFpcore(self, ctx):
        input_vars = [t.text for t in ctx.inputs]
        props = {name: x for name, x in (p.accept(self) for p in ctx.props)}
        e = ctx.e.accept(self)
        return FPCoreObject(input_vars, props, e)

    def visitFpimp(self, ctx):
        raise ValueError('unsupported: FPImp')

    def visitExprNum(self, ctx):
        #print('Num: ' + ctx.c.text)
        return ast.Val(ctx.c.text)

    def visitExprConst(self, ctx):
        return ast.Val(ctx.c.text) # this won't work with numpy...

    def visitExprVar(self, ctx):
        #print('Var: ' + ctx.x.text)
        return ast.Var(ctx.x.text)

    def visitExprUnop(self, ctx):
        op = ctx.op.getText()
        # special case unary negation
        if op == '-':
            return ops['neg'](ctx.arg0.accept(self))
        else:
            return ops[op](ctx.arg0.accept(self))

    def visitExprBinop(self, ctx):
        op = ctx.op.getText()
        return ops[op](ctx.arg0.accept(self), ctx.arg1.accept(self))

    def visitExprComp(self, ctx):
        op = ctx.op.getText()
        return ops[op](*(e.accept(self) for e in ctx.args))

    def visitExprLogical(self, ctx):
        op = ctx.op.getText()
        return ops[op](*(e.accept(self) for e in ctx.args))

    def visitExprIf(self, ctx):
        raise ValueError('unsupported: If')

    def visitExprLet(self, ctx):
        raise ValueError('unsupported: Let')

    def visitExprWhile(self, ctx):
        raise ValueError('unsupported: While')

    def visitPropStr(self, ctx):
        return ctx.name.text, ctx.s.text[1:-1]

    def visitPropList(self, ctx):
        return ctx.name.text, [t.text for t in ctx.syms]

    def visitPropExpr(self, ctx):
        return ctx.name.text, ctx.e.accept(self)

def parse(s):
    input_stream = antlr4.InputStream(s)
    lexer = FPCoreLexer(input_stream)
    token_stream = antlr4.CommonTokenStream(lexer)
    parser = FPCoreParser(token_stream)
    parse_tree = parser.parse()
    return parser, parse_tree

def compile(s):
    parser, tree = parse(s)
    visitor = Visitor()
    return visitor.visit(tree)

if __name__ == '__main__':
    import sys
    parser, tree = parse(sys.stdin.read())

    visitor = Visitor()
    results = visitor.visit(tree)
    for x in results:
        print(str(x))
        print()
        print(repr(x))
        print()
        print(x.sexp)

    #print(results[0].e({'x': 0.125}));
        

File: www/fpcparser.js
const $ = require("jquery");
const antlr4 = require("antlr4");
const FPCoreLexer = require("./gen/FPCoreLexer.js").FPCoreLexer;
const FPCoreParser = require("./gen/FPCoreParser.js").FPCoreParser;
const FPCoreVisitor = require("./gen/FPCoreVisitor.js").FPCoreVisitor;
const ast = require("./fpcast.js");
const ops = ast.operations;

class Visitor extends FPCoreVisitor {
    visitParse(ctx) {
        const cores = [];
        for (const child of ctx.children) {
            const parsed = child.accept(this);
            if (parsed) {
                cores.push(parsed);
            }
        }
        return cores;
    }

    visitFpcore(ctx) {
        const input_vars = ctx.inputs.map(x => x.text);
        const props = {};
        for (const child of ctx.props) {
            const [name, x] = child.accept(this);
            props[name] = x;
        }
        const e = ctx.e.accept(this);
        const core = {
            args: input_vars,
            properties: props,
            name: props[":name"],
            pre: props[":pre"],
            e: e,
        };
        return core;
    }

    visitFpimp(ctx) {
        throw new Error("unsupported: FPImp");
    }

    visitExprNum(ctx) {
        return new ast.Val(ctx.c.text);
    }

    visitExprConst(ctx) {
        return new ast.Val(ctx.c.text);
    }

    visitExprVar(ctx) {
        return new ast.Var(ctx.x.text);
    }

    visitExprUnop(ctx) {
        const op = ctx.op.getText();
        if (op === "-") {
            return new ops["neg"](ctx.arg0.accept(this));
        } else {
            return new ops[op](ctx.arg0.accept(this));
        }
    }

    visitExprBinop(ctx) {
        const op = ctx.op.getText();
        return new ops[op](ctx.arg0.accept(this), ctx.arg1.accept(this));
    }

    visitExprComp(ctx) {
        const op = ctx.op.getText();
        const visitor = this;
        return new ops[op](...(function*() {
            for (const e of ctx.args) {
                yield e.accept(visitor);
            }
        })());
    }

    visitExprLogical(ctx) {
        const op = ctx.op.getText();
        const visitor = this;
        return new ops[op](...(function*() {
            for (const e of ctx.args) {
                yield e.accept(visitor);
            }
        })());
    }

    visitExprIf(ctx) {
        throw new Error("unsupported: If");
    }

    visitExprLet(ctx) {
        throw new Error("unsupported: Let");
    }

    visitExprWhile(ctx) {
        throw new Error("unsupported: While");
    }

    visitPropStr(ctx) {
        return [ctx.name.text, ctx.s.text.slice(1, -1)];
    }

    visitPropList(ctx) {
        return [ctx.name.text, ctx.syms.map(x => x.text)];
    }

    visitPropExpr(ctx) {
        return [ctx.name.text, ctx.e.accept(this)];
    }
}

function parse(input) {
    const input_stream = new antlr4.InputStream(input);
    const lexer = new FPCoreLexer(input_stream);
    const token_stream  = new antlr4.CommonTokenStream(lexer);
    const parser = new FPCoreParser(token_stream);
    const parse_tree = parser.parse();
    return [parser, parse_tree];
}

function compile(input) {
    const [parser, tree] = parse(input);
    visitor = new Visitor();
    return visitor.visit(tree);
}

// web interface

function onCore() {
    const root = parentForm(this);
    const userCore = $(".incore", root).val();
    if (userCore) {
        const compiledCore = compile(userCore)[0];
        let coreStr = compiledCore.e.toString();
        if (compiledCore.pre) {
            coreStr = "pre: " + compiledCore.pre.toString() + "\n" + coreStr;
        }
        coreStr = "FPCore (" + compiledCore.args.join(" ") + ")\n" + coreStr;
        $(".outcore", root).html(coreStr);
        root.data("core", compiledCore);
        onArgs.call(this);
    }
}

function onArgs() {
    const root = parentForm(this);
    const compiledCore = root.data("core");
    if (compiledCore) {
        const userArgs = $(".arguments", root).val();
        const args = userArgs.split(";").map(x => x.trim());
        const ctx = {};
        for (let i = 0; i < compiledCore.args.length; i++) {
            ctx[compiledCore.args[i]] = args[i];
        }
        let resultStr = "result: " + compiledCore.e.apply(ctx);
        if (compiledCore.pre) {
            resultStr = "pre: " + compiledCore.pre.apply(ctx) + "\n" + resultStr;
        }
        $(".outval", root).html(resultStr);
    }
}

// set up the web interface

function parentForm(element) {
    return $(element).closest(".fpc");
}

function setupCore(rootClass) {
    const root = $(rootClass);
    $(".incore", root).on("input", onCore);
    $(".arguments", root).on("input", onArgs);
    onCore.call($(".incore", root).get(0));
}

exports.setupCore = setupCore;


//testing
// let inputDoc = "";

// process.stdin.setEncoding("utf8");

// process.stdin.on("readable", () => {
//     const chunk = process.stdin.read();
//     if (chunk) {
//         inputDoc += chunk;
//     }
// });

// process.stdin.on("end", () => {
//     const cores = compile(inputDoc);
//     console.log(cores[0]);
//     console.log(cores[0].e.apply({"x": 0.125}));
// });
